### PR TITLE
feat: enable decoding non-Ethereum precompiles when tracing

### PR DESCRIPTION
**Pull request by 4meta5**
Closes https://github.com/foundry-rs/foundry/issues/1078

Enable decoding calls from non-Ethereum precompiles when tracing by adding a field to `foundry_config::Config` for `precompiles: Vec<(Address, Function)>`.

## Motivation

Enable decoding calls from non Ethereum precompiles when tracing https://github.com/foundry-rs/foundry/issues/1078

## Solution

Define `with_precompiles` method on `CallTraceDecoderBuilder` to augment the Ethereum precompiles included in `new` method  with non-Ethereum precompiles specified by a config.

### CODE REVIEW

File: config/src/lib.rs [hunk 1]
```diff
@@ -197,6 +198,9 @@ pub struct Config {
     pub model_checker: Option<ModelCheckerSettings>,
     /// verbosity to use
     pub verbosity: u8,
+    /// Information for decoding custom precompile calls
+    #[serde(default)]
+    pub precompiles: Vec<(Address, Function)>,
```
Comment (mattsse) [inline config/src/lib.rs:203]: this is reasonable,

I'd like to see some tests/examples for this in toml format

**Review by mattsse — CHANGES_REQUESTED** [review]

File: evm/src/trace/decoder.rs [hunk 2]
```diff
@@ -51,6 +51,12 @@ impl CallTraceDecoderBuilder {
         self
     }
 
+    /// Add information for decoding custom precompile calls.
+    pub fn with_precompiles(mut self, precompiles: HashMap<Address, Function>) -> Self {
+        self.decoder.precompiles = self.decoder.precompiles.into_iter().chain(precompiles).collect();
```
Comment (0xMelkor) [inline evm/src/trace/decoder.rs:56]: I'd keep `CallTraceDecoder`'s  encapsulation safe. 
How if: 
```rust
impl CallTraceDecoder {
  pub fn chain(&mut self, precompiles: HashMap<Address, Function>) {
     ...
  }
}
```

**Review by Evalir — APPROVED** [review]
should be good now—just needs examples

**Comment by Evalir** [comment]
Will close this—while "complete", it's a pretty sensitive change that will get overwritten by the alloy migration, and on top of this we're still debating a bit on how to include these external precompiles. Once the migration is done we can include a configuration option to detect and decode these.

**Pull request closed without merging**